### PR TITLE
Add translation rule for optimization barrier.

### PR DESCRIPTION
**Pull request by pschuh**
Also adds a translation rule for remat that uses the new optimization barrier
op. If you find errors, consider disabling the remat lowering using
`jax_remat_opt_barrier` config flag.

### CODE REVIEW

File: jax/_src/lax/control_flow.py [hunk 1]
```diff
@@ -3122,4 +3130,39 @@ def _remat_translation_rule(*args,
...
+  return args
+
+
+def _optimization_barrier_translation_rule(ctx, avals_in, avals_out, *args):
+  out = xops.OptimizationBarrier(xops.Tuple(ctx.builder, args))
+  return [xops.GetTupleElement(out, i) for i in range(len(args))]
+
+
+def _optimization_barrier_lowering_rule(ctx, *args):
+  loop_carry_types = _map(mlir.aval_to_ir_types, ctx.avals_in)
```
Comment (hawkinsp) [inline jax/_src/lax/control_flow.py:3147]: I suspect `loop_carry` is a poor choice of name here, given there are no longer any loops.

File: jax/_src/lax/control_flow.py [hunk 2]
```diff
@@ -3122,4 +3130,39 @@ def _remat_translation_rule(*args,
...
+  flat_loop_carry_types = util.flatten(loop_carry_types)
+
+  flat_args = mlir.flatten_lowering_ir_args(args)
+  barrier_op = mhlo.OptimizationBarrierOp(flat_loop_carry_types, flat_args)
```
Comment (hawkinsp) [inline jax/_src/lax/control_flow.py:3151]: BTW, you can avoid the need to provide the output type if you implement a type inference rule on the MHLO op. You'd need to do that separately in another change, of course.

**Review by hawkinsp — APPROVED** [review]

File: jax/_src/lax/control_flow.py (hunk 1 above)
Comment (pschuh) [reply jax/_src/lax/control_flow.py:3147]: Done.

File: jax/_src/lax/control_flow.py (hunk 2 above)
Comment (pschuh) [reply jax/_src/lax/control_flow.py:3151]: Ah, good to know. I was curious why this didn't just work.

**Review by hawkinsp — APPROVED** [review]

**Pull request merged**